src/abi.md

```diff
@@ -1,21 +1,246 @@
 # Application Binary Interface (ABI)
 
+r[abi]
+
 This section documents features that affect the ABI of the compiled output of
 a crate.
 
 See *[extern functions]* for information on specifying the ABI for exporting
-functions. See *[external blocks]* for information on specifying the ABI for
+functions. See *[`extern` block]s* for information on specifying the ABI for
 linking external libraries.
 
+## ABI compatibility
+
+r[abi.compatibility]
+
+r[abi.compatibility.intro]
+Function calls pass parameters and return values between the caller and the callee function. This requires the caller and callee to agree on an ABI for those parameters and return values. This is typically only guaranteed when the same type is used in both the call site and the definition site of the callee, but certain other types may be *abi compatible*. This can appear when transmuting a [function pointer] or using an [`extern` block] to call a function. When the parameters or return types differ between the call site and def site, assuming they are *abi compatible*, the parameters or return types are transmuted to the type at the def site or call site respectively.
+
+> [!NOTE]
+> This can include calls to functions defined outside of rust, or built using a different Rust compiler version.
+> Additional guarantees will apply in this case for "FFI Safe" types, which match up with the platform C ABI in well-defined ways.
+> These are not fully documented here currently.
```

Comment:
Shouldn't this say that additional *requirements* apply? The document right now only describes the rules for Rust-to-Rust calls, as far as I can see.

Reply:
No, because the note is talking about the fact that "FFI Safe" types have a defined mapping to the platform C abi. 
You can technically pass a non-"FFI Safe" type to an FFI call, there's just no guarantee you can match it on the other side.